### ISSUE TITLE
Indentation broken with multiline literal initializer expressions

### ISSUE THREAD

**Issue by Nikratio** [microsoft/vscode-cpptools#9230]
Bug type: Language Service

On Linux with VSCode 1.66.2 and C/C++ Extension Pack 1.1.0, I'm seeing the following:

```
char **argv = {"foo",<enter>
|
```
i.e., pressing enter after the first line puts the cursor in column 1, without any indentation. This is wrong, an open brace should result in at least some level of indentation.

Moreover, after manually indenting:

```
char **argv = {"foo", 
   NULL};<enter>
  |
```
i.e, now the cursor is indented at the wrong level (since there no longer is any open brace).

Note that selecting the area and formatting it with "Format Selection" gives the correct result.

**Comment by Nikratio** [comment]
Thanks for looking into this!

I have not changed anything from the default settings. Formatting on enter seems to work correctly in other cases (eg. after `if (1) {<enter>` I'm getting indentation).

Could you provide more details on what variables should be set to what values? 

When you say " the behavior you want", do you mean that some users *want* to have (non-)indentation like

```
char **argv = {"foo", 
NULL};
```
? If so, I have a hard time believing this...

**Comment by Colengms** [comment]
>I have not changed anything from the default settings. Formatting on enter seems to work correctly in other cases

Hi @Nikratio .  As I mentioned, the default settings for cpptools (using clang-format) do not trigger formatting/indenting when enter is pressed.  The indenting you are seeing is a result of behavior from VS Code itself, based on settings such as: `editor.autoIndent`, `editor.detectIndentation`.  I believe that is what you're seeing when indenting occurs for `if (1) {<enter>`.  You should see the same behavior with the C/C++ extension disabled.

This behavior from VS Code can actually conflict or change the apparent behavior of similar functionality when using `vcFormat` formatting from the C/C++ extension, as VS Code tries to apply the results of both features.

Since it sounds like you are not configured to get on-type formatting/indenting from the C/C++ extension, perhaps the issue you are seeing is actually specific to VS Code itself?  If you think VS Code should be honoring that open curly brace and indenting when you press enter, consider opening an issue in [the VS Code repo](https://github.com/microsoft/vscode/issues).

> Could you provide more details on what variables should be set to what values?

I was mainly referring to the need to potentially choose between `vcFormat` indenting settings or similar settings provided by VS Code, as having both occur may not work as expected.  Otherwise, how your code is formatted is up to you.  Documentation for the `vcFormat` related settings are available in the VS Code settings UI, under Extensions -> C/C++, under `C_Cpp > Vc Format > *`.  Using `vcFormat` as the formatting engine also requires setting `C_Cpp.formatting` to `vcFormat`.  I suggest vcFormat as an alternative, as it provides indenting features that may work for you.

**Comment by Nikratio** [comment]
Thanks for your help! I have now set ` "C_Cpp.formatting": "vcFormat"`, which I think means that vcFormat will set the indentation on enter?

The problem is still the same as in my first message. However, now explicit formatting with Ctrl-K Ctrl-F does not indent the code either.

Looking at potentially related settings, the only thing that I found is:

```
"C_Cpp.vcFormat.indent.withinParentheses": "indent"
"C_Cpp.vcFormat.indent.multiLineRelativeTo": "innermostParenthesis"
"C_Cpp.vcFormat.indent.preserveWithinParentheses": false
```

Which seems correct..  am I missing something still?

**Comment by Colengms** [comment]
@Nikratio Could you provide an example that does not format properly when you explitly format it?  The following worked fine for me:

```
int foo()
{
if (1)
{
int i;
}
}
```

```
int foo()
{
    if (1)
    {
        int i;
    }
}
```

Using the following settings:
```
    "C_Cpp.formatting": "vcFormat",
    "C_Cpp.vcFormat.indent.withinParentheses": "indent",
    "C_Cpp.vcFormat.indent.multiLineRelativeTo": "innermostParenthesis",
    "C_Cpp.vcFormat.indent.preserveWithinParentheses": false
```

Note that the behavior of vcFormat when typing Enter is also influenced by any of the `C_Cpp.vcFormat.newline.*` settings.

**Comment by Nikratio** [comment]
Sure! It's in my first message:

```
char **argv = { "foo",
"bar",
NULL };
```

**Comment by Colengms** [comment]
Hi @Nikratio .  With those settings I'm seeing indentation occur when a new line is typed after a parenthesis, brace, bracket, or angle brace.  Could you confirm you have `"editor.formatOnType": true` set?

You're right about it not applying indentation to an initializer list when explicitly formatting.  Although vcFormat appears to apply the indent-on-type rule for an initializer list (as well as for angle braces, brackets, etc), it looks like that behavior is not influenced by the `withinParentheses` or `preserveWithinParentheses` settings (perhaps because they are not technically parentheses).

I've opened an internal issue against VS asking for these others scenarios to either be handled by the existing settings, or to add new settings to allow their indent behavior to be modified.

**Comment by Nikratio** [comment]
Yes, `formatOnType` is set. I am also seeing correct indentation when pressing enter after a brace. The problem is probably that in my example, the line does not end in a brace. Still, I would expect VS Code to get that right.

Thanks for following up on this!
